### DISCUSSION TITLE
Button: new note based on template

### DISCUSSION THREAD

**Discussion by glen-w** [TriliumNext/Trilium#2309] [Q&A] [answered]
I want to add a button that creates a new note using an existing template. I thought this should be simple enough but it is my first foray into coding in Trilium (with only limited javascript experience generally) and I am really struggling. 

I have used the example script for creating a button and am trying to add the 'template' attribute to the newly created using CreateNoteAttribute. The note gets created but the template is not applied. Any help would be greatly appreciated!

```
api.addButtonToToolbar({
    title: 'log interaction',
    icon: 'user',
    shortcut: 'alt+i',
    action: async () => {

        const newNoteId = await api.runOnServer(() => {
            const todayNote = api.getTodayNote();
            const resp = api.createTextNote(todayNote.noteId, 'new interaction', '');
            
           //add template attribute to new note
            resp.api.CreateNoteAttribute('relation','template','log interaction template');
            
            return resp.note.newNoteId;
            
        });
        
        // wait until the frontend is fully synced with the changes made on the backend above
        await api.waitUntilSynced();

        // we got an ID of newly created note and we want to immediatelly display it
        await api.activateNewNote(newNoteId);
    }
});
```

**Comment by MorganeLeCaignec** [comment, marked as answer]
Hello! 

Basically your problem is that adding a template to a note means adding a **relation**, linking your new note to another existing note with a relation attribute. For that, you need .setRelation with the template's noteId.
Right now in your example, you're creating an attribute of type 'relation' to nothing, because 'template' is just a meaningless string, it's not what a relation attribute requires.

I've done something similar in my Hexmap widget and here's how, modifying your example:

```
api.addButtonToToolbar({
    title: 'log interaction',
    icon: 'user',
    shortcut: 'alt+i',
    action: async () => {

        const newNote = await api.runOnServer(() => {
            const todayNote = api.getTodayNote();
            const newNoteObject = api.createTextNote(todayNote.noteId, 'new interaction', '');
            
            // You need either the template's noteId, or you get the template note itself in some other way (which I prefer).
            // To get the template note, here I grab it from a specific unique label attribute that exists only on your template, you could also get it from a note's relation attribute to the template (so you get a note, that note is linked to the template through a relation attribute, you grab the relation attribute).
            const templateNote = api.getNotesWithLabel('your_unique_template_label')[0];
            //add template to new note
            newNoteObject.note.setRelation('template', templateNote.noteId);
            
            return newNoteObject.note;
        });
        
        // wait until the frontend is fully synced with the changes made on the backend above
        await api.waitUntilSynced();

        // we got an ID of newly created note and we want to immediatelly display it
        await api.activateNewNote(newNote.noteId);
    }
});
```

I've changed a few variables names and such when I was testing because it felt more consistent.
This method, getting a note from a unique label attribute I've decided on, or through my main render/widget note's relation attribute (which means it's easily customizable) works well for me. Keep in mind this doesn't check if the template note exists or not (if you typed your label wrong, removed it, etc..) so I'd add an if check to make sure templateNote exists/is not empty.

If you're interested, my two widgets (Hexmap widget and Timeline widget) are on my [gitlab](https://gitlab.com/QuentinLeCaignec) with the source code, you can also download the .zip and import it into your own install to test them and see what they do. If you're not already doing it, I also recommend you open the browser console (because trilium is an electron app, meaning it's built on chromium and functions basically like google chrome) with CTRL+SHIFT+i, so you can watch the console for errors and better understand what's going wrong.

**Reply by glen-w** [reply]
Thanks so much Quentin, you made my day! This is exactly what I needed and your clear explanations are super helpful. I didn't know about Electron apps - great tip, thanks.

Have now replicated for other templates I want easy access to:

<img width="178" alt="Capture d’écran 2021-11-06 à 12 19 54" src="https://user-images.githubusercontent.com/6459622/140607725-47a1d765-3cf8-4b91-b399-54802b4fc5f5.png">

**Reply by MorganeLeCaignec** [reply]
Glad it works! :+1:
